src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs

```diff
@@ -49,6 +87,96 @@ private Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs a
                 return null;
             }
         }
+
+        private AssemblyIdentity? GetOrAddAssemblyIdentity(string fullPath)
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+
+            lock (_guard)
+            {
+                if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity))
+                {
+                    return existingIdentity;
+                }
+            }
+
+            var identity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
+            return AddToCache(fullPath, identity);
+        }
+
+        [return: NotNullIfNotNull("identity")]
+        private AssemblyIdentity? AddToCache(string fullPath, AssemblyIdentity? identity)
+        {
+            lock (_guard)
+            {
+                if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity) && existingIdentity != null)
+                {
+                    identity = existingIdentity;
+                }
+                else
+                {
+                    _loadedAssemblyIdentitiesByPath[fullPath] = identity;
+                }
+            }
+
+            return identity;
+        }
+
+        private Assembly? Load(string displayName)
+        {
+            if (!AssemblyIdentity.TryParseDisplayName(displayName, out var requestedIdentity))
+            {
+                return null;
+            }
+
+            ImmutableArray<string> candidatePaths;
+            lock (_guard)
+            {
+
+                // First, check if this loader already loaded the requested assembly:
+                if (_loadedAssembliesByIdentity.TryGetValue(requestedIdentity, out var existingAssembly))
+                {
+                    return existingAssembly;
+                }
+                // Second, check if an assembly file of the same simple name was registered with the loader:
+                var pathList = GetPaths(requestedIdentity.Name);
+                if (pathList is null)
+                {
+                    return null;
+                }
+
+                Debug.Assert(pathList.Count > 0);
+                candidatePaths = pathList.ToImmutableArray();
+            }
+
+            // Find the highest version that satisfies the original request. We'll match for the highest version we can, assuming it
+            // actually matches the original request
+            string? bestPath = null;
+            Version? bestIdentityVersion = null;
+
+            foreach (var candidatePath in candidatePaths)
```

Comment:
This introduces a bit of non-determinism. Consider that if the same assembly is listed with two different paths the path we pick for the assembly is non-deterministic because we don't sort the paths here. 

That will matter because while it's identical assemblies the contents of the containing directory are not necessarily identical. If memory serves the CLR will look for dependencies of an assembly in `LoadFrom` in the same directory the assembly was loaded from. Hence the set of dependent `Assembly` resolved could be diff in diff directories.

Reply:
@jaredpar any specific preference to the preferred sort, or just sort via StringComparer.Ordinal or something?

Oddly the current comparer for the hashset-based comparer is PathComparer, which does a bunch of extra comparisons around relative pathing and such to attempt to "normalize" the path. I was going to just switch this over to an ImmutableSortedSet but PathComparer doesn't implement that (since it would seem insane to do so).

Reply:
Lets use `StringComparer.Ordinal`. It's really the only sane decision when it comes to path analysis

Reply:
@jaredpar Switched to sort with StringComparer.Ordinal; there's still a different equality comparer in the hash set that is used for determining whether two paths are the same, that's being left untouched (but would be an easy follow-up PR to clean up.)